p2p/transport_memory.go

```diff
@@ -0,0 +1,394 @@
+	case <-peerTransport.closeCh:
+		return nil, ErrTransportClosed{}
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+// DialAccept is a convenience function that dials a peer MemoryTransport and
+// returns both ends of the connection (A to B and B to A).
+func (t *MemoryTransport) DialAccept(
+	ctx context.Context,
+	peer *MemoryTransport,
+) (Connection, Connection, error) {
+	endpoints := peer.Endpoints()
+	if len(endpoints) == 0 {
+		return nil, nil, fmt.Errorf("peer %q not listening on any endpoints", peer.nodeInfo.DefaultNodeID)
+	}
+
+	acceptCh := make(chan Connection, 1)
+	errCh := make(chan error, 1)
+	go func() {
+		conn, err := peer.Accept(ctx)
+		errCh <- err
+		acceptCh <- conn
+	}()
+
+	outConn, err := t.Dial(ctx, endpoints[0])
+	if err != nil {
+		return nil, nil, err
+	}
+	if err = <-errCh; err != nil {
+		return nil, nil, err
+	}
+	inConn := <-acceptCh
+
+	return outConn, inConn, nil
+}
+
+// Close implements Transport.
+func (t *MemoryTransport) Close() error {
+	err := t.network.RemoveTransport(t.nodeInfo.DefaultNodeID)
+	t.closeOnce.Do(func() {
+		close(t.closeCh)
+	})
+	t.logger.Info("stopped accepting connections")
+	return err
+}
+
+// Endpoints implements Transport.
+func (t *MemoryTransport) Endpoints() []Endpoint {
+	select {
+	case <-t.closeCh:
+		return []Endpoint{}
+	default:
+		return []Endpoint{{
+			Protocol: MemoryProtocol,
+			PeerID:   t.nodeInfo.DefaultNodeID,
+			Path:     string(t.nodeInfo.DefaultNodeID),
+		}}
+	}
+}
+
+// SetChannelDescriptors implements Transport.
+func (t *MemoryTransport) SetChannelDescriptors(chDescs []*conn.ChannelDescriptor) {
+}
+
+// MemoryConnection is an in-memory connection between two transports (nodes).
+type MemoryConnection struct {
+	logger log.Logger
+	local  *MemoryTransport
+	remote *MemoryTransport
+
+	receiveCh <-chan memoryMessage
+	sendCh    chan<- memoryMessage
+	closeCh   <-chan struct{}
+	close     func() bool
+}
+
+// memoryMessage is used to pass messages internally in the connection.
+type memoryMessage struct {
+	channel byte
+	message []byte
+}
+
+// newMemoryConnection creates a new MemoryConnection. It takes all channels
+// (including the closeCh signal channel) on construction, such that they can be
+// shared between both ends of the connection.
+func newMemoryConnection(
+	local *MemoryTransport,
+	remote *MemoryTransport,
+	receiveCh <-chan memoryMessage,
+	sendCh chan<- memoryMessage,
+	closeCh <-chan struct{},
+	close func() bool,
+) *MemoryConnection {
+	c := &MemoryConnection{
+		local:     local,
+		remote:    remote,
+		receiveCh: receiveCh,
+		sendCh:    sendCh,
+		closeCh:   closeCh,
+		close:     close,
+	}
+	c.logger = c.local.logger.With("remote", c.RemoteEndpoint())
+	return c
+}
+
+// ReceiveMessage implements Connection.
+func (c *MemoryConnection) ReceiveMessage() (chID byte, msg []byte, err error) {
+	// check close first, since channels are buffered
+	select {
+	case <-c.closeCh:
+		return 0, nil, io.EOF
+	default:
+	}
```

Comment:
Why do we need to do this? The go `c.close` can be closed between this block and the next `select`.

Reply:
This is to make sure that if the connection is _already_ closed, we never receive or send a message. For example, if `receiveCh` contains a buffered message but the connection has been closed, without this check there is a 50% chance that the buffered message will be returned even though the connection has been closed.

If the connection is closed after this check (concurrently with the `select`) there is also a 50% chance that the message will be returned, but that is consistent with concurrency semantics where either the receive event or the close event may happen first depending on goroutine/channel scheduling.

Reply:
That makes sense, but I don't see what first select buys us since there is still a chance we receive and return a message if the connection is closed before the last select.

Reply:
Without the select, a single (non-concurrent) client can see this behavior:

```
connection.Close()
msg := connection.ReceiveMessage() // returns a message!
```

This is clearly wrong. If they are concurrent, however, it is entirely reasonable for there to be a 50% chance of receiving a message.

```
go func() {
  connection.Close()
}()

go func() {
  msg := connect.ReceiveMessage() // 50% chance of returning, entirely reasonable
}()
```

The first select is necessary to make connection closures linearizable, i.e. once the connection is closed it stays closed.

Reply:
Ok, I think that's reasonable 👍